helpers/pc-bios.c

```diff
@@ -0,0 +1,50 @@
+/*
+ * Lightweight ACPI Implementation
+ * Copyright (C) 2018-2019 the lai authors
+ */
+
+#include <acpispec/tables.h>
+#include <lai/core.h>
+#include <lai/helpers/pc-bios.h>
+
+static uint8_t lai_bios_calc_checksum(void *ptr, size_t size) {
+	uint8_t sum = 0;
+	for (size_t i = 0; i < size; i++)
+		sum += ((uint8_t *)ptr)[i];
+	return sum;
+}
+
+int lai_bios_detect_rsdp_within(uintptr_t base, size_t length,
+		uintptr_t *rsdp_addr, uintptr_t *rsdt_addr) {
+    int version = 0;
+	uint8_t *window = laihost_map(base, length);
+	for (size_t off = 0; off < length; off += 16) {
+		acpi_rsdp_t *rsdp = (acpi_rsdp_t *)(window + off);
+
+		if (memcmp(rsdp->signature, "RSD PTR ", 8))
+			continue;
+		if (lai_bios_calc_checksum(rsdp, sizeof(acpi_rsdp_t)))
+			continue;
+
+        // TODO: Handle XSDTs.
+		if(rsdp_addr)
+			*rsdp_addr = base + off;
+		if(rsdt_addr)
+			*rsdt_addr = rsdp->rsdt;
+		version = 1;
+        goto done;
+	}
+
+done:
+    // TODO: Unmap the memory range.
+	return version;
+}
+
+int lai_bios_detect_rsdp(uintptr_t *rsdp_addr, uintptr_t *rsdt_addr) {
+	int version;
+	if((version = lai_bios_detect_rsdp_within(0x80000, 0x20000, rsdp_addr, rsdt_addr)))
```

Comment:
Technically the EBDA is not at `0x80000` for every PC, however the bios does specify the address of it in the BDA (see https://wiki.osdev.org/EBDA#BIOS_Data_Area_.28BDA.29 especially 0x40E). However to not open lai up to bios bugs some simply validity checking would have to be done

Reply:
You're right. ACPI does specify 0x40E though, so that should be more or less reliable, at least on hardware that supports ACPI.